atecc608a_se.c

```diff
@@ -0,0 +1,224 @@
+    ASSERT_STATUS(atcab_init(&atca_iface_config), ATCA_SUCCESS);
+    ASSERT_STATUS(atcab_read_serial_number(serial), ATCA_SUCCESS);
+    ASSERT_STATUS(atcab_release(), ATCA_SUCCESS);
+    printf("Serial Number:\n");
+    atcab_printbin_sp(serial, ATCA_SERIAL_NUM_SIZE);
+    printf("\n");
+
+    return 0;
+}
+
+psa_status_t atecc608a_to_psa_error(ATCA_STATUS ret)
+{
+    switch (ret)
+    {
+    case ATCA_SUCCESS:
+    case ATCA_RX_NO_RESPONSE:
+    case ATCA_WAKE_SUCCESS:
+        return PSA_SUCCESS;
+    case ATCA_BAD_PARAM:
+    case ATCA_INVALID_ID:
+    case ATCA_INVALID_SIZE:
+    case ATCA_SMALL_BUFFER:
+    case ATCA_BAD_OPCODE:
+    case ATCA_ASSERT_FAILURE:
+        return PSA_ERROR_INVALID_ARGUMENT;
+    case ATCA_RX_CRC_ERROR:
+    case ATCA_RX_FAIL:
+    case ATCA_STATUS_CRC:
+    case ATCA_RESYNC_WITH_WAKEUP:
+    case ATCA_PARITY_ERROR:
+    case ATCA_TX_TIMEOUT:
+    case ATCA_RX_TIMEOUT:
+    case ATCA_TOO_MANY_COMM_RETRIES:
+    case ATCA_COMM_FAIL:
+    case ATCA_TIMEOUT:
+    case ATCA_TX_FAIL:
+    case ATCA_NO_DEVICES:
+        return PSA_ERROR_COMMUNICATION_FAILURE;
+    case ATCA_UNIMPLEMENTED:
+        return PSA_ERROR_NOT_SUPPORTED;
+    case ATCA_ALLOC_FAILURE:
+        return PSA_ERROR_INSUFFICIENT_MEMORY;
+    case ATCA_CONFIG_ZONE_LOCKED:
+    case ATCA_DATA_ZONE_LOCKED:
+    case ATCA_NOT_LOCKED:
+    case ATCA_WAKE_FAILED:
+    case ATCA_STATUS_UNKNOWN:
+    case ATCA_STATUS_ECC:
+    case ATCA_STATUS_SELFTEST_ERROR:
+    case ATCA_CHECKMAC_VERIFY_FAILED:
+    case ATCA_PARSE_ERROR:
+    case ATCA_FUNC_FAIL:
+    case ATCA_GEN_FAIL:
+    case ATCA_EXECUTION_ERROR:
+    case ATCA_HEALTH_TEST_ERROR:
+    default:
+        return PSA_ERROR_HARDWARE_FAILURE;
+    }
+}
+
+ATCA_STATUS checkLockedZones()
+{
+    bool locked;
+    bool config_locked;
+    printf("--- Device locks information ---\n");
+    ASSERT_STATUS(atcab_init(&atca_iface_config), ATCA_SUCCESS);
+    ASSERT_STATUS(atcab_is_locked(LOCK_ZONE_CONFIG, &config_locked), ATCA_SUCCESS);
+    printf("  - Config locked: %d\n", config_locked);
+    ASSERT_STATUS(atcab_is_locked(LOCK_ZONE_DATA, &locked), ATCA_SUCCESS);
+    printf("  - Data locked: %d\n", locked);
+    for(uint8_t i=0; i < 16; i++)
+    {
+        ASSERT_STATUS(atcab_is_slot_locked(i, &locked), ATCA_SUCCESS);
+        printf("  - Slot %d locked: %d\n", i, locked);
+    }
+    ASSERT_STATUS(atcab_release(), ATCA_SUCCESS);
+    printf("--------------------------------\n");
+    if(!config_locked)
+    {
+        printf("Config is not locked! Cannot use private keys.\n");
+        return ATCA_NOT_LOCKED;
+    }
+    return ATCA_CONFIG_ZONE_LOCKED;
+}
+
+psa_status_t atecc608a_export_key(psa_key_slot_number_t key,
+                                         uint8_t *p_data,
+                                         size_t data_size,
+                                         size_t *p_data_length)
+{
+    ATCA_STATUS ret = ATCA_SUCCESS;
+    uint16_t slot = key;
+    size_t key_data_len = 65;
```

Comment:
This function does not implement the right format. A private EC key is represented as just the private value, a 32-byte string for a 256-bit curve. The public key is not stored in the representation of the private key.

Reply:
This is the private key exporting function, I just got the name wrong.